File: src/main/kotlin/com/huawei/cangjie/psi/stubs/elements/CjAnnotationEntryElementType.java
package com.huawei.cangjie.psi.stubs.elements;

import com.huawei.cangjie.name.Name;
import com.huawei.cangjie.psi.CjAnnotationEntry;
import com.huawei.cangjie.psi.CjValueArgumentList;
import com.huawei.cangjie.psi.stubs.CangJieAnnotationEntryStub;
import com.huawei.cangjie.psi.stubs.impl.CangJieAnnotationEntryStubImpl;
import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.StringRef;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public class CjAnnotationEntryElementType extends CjStubElementType<CangJieAnnotationEntryStub, CjAnnotationEntry>{

    public CjAnnotationEntryElementType(@NotNull @NonNls String debugName) {
        super(debugName, CjAnnotationEntry.class, CangJieAnnotationEntryStub.class);
    }

    @Override
    public @NotNull CangJieAnnotationEntryStub createStub(@NotNull CjAnnotationEntry psi, StubElement<? extends PsiElement> parentStub) {
        Name shortName = psi.getShortName();
        String resultName = shortName != null ? shortName.asString() : null;
        CjValueArgumentList valueArgumentList = psi.getValueArgumentList();
        boolean hasValueArguments = valueArgumentList != null && !valueArgumentList.getArguments().isEmpty();
        return new CangJieAnnotationEntryStubImpl((StubElement<?>) parentStub, StringRef.fromString(resultName), hasValueArguments );

    }

    @Override
    public void serialize(@NotNull CangJieAnnotationEntryStub stub, @NotNull StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getShortName());
        dataStream.writeBoolean(stub.hasValueArguments());
        if (stub instanceof CangJieAnnotationEntryStubImpl) {
//            Map<Name, ConstantValue<?>> arguments = ((CangJieAnnotationEntryStubImpl) stub).getValueArguments();
//            dataStream.writeInt(arguments != null ? arguments.size() : 0);
//            if (arguments != null) {
//                for (Map.Entry<Name, ConstantValue<?>> valueEntry : arguments.entrySet()) {
//                    dataStream.writeName(valueEntry.getKey().asString());
//                    ConstantValue<?> value = valueEntry.getValue();
//                    CangJieConstantValueKt.serialize(value, dataStream);
//                }
//            }
        }
    }

    @Override
    public @NotNull CangJieAnnotationEntryStub deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        StringRef text = dataStream.readName();
        boolean hasValueArguments = dataStream.readBoolean();
//        int valueArgCount = dataStream.readInt();
//        Map<Name, ConstantValue<?>> args = new LinkedHashMap<>();
//        for (int i = 0; i < valueArgCount; i++) {
//            args.put(Name.identifier(Objects.requireNonNull(dataStream.readNameString())),
//                    KotlinConstantValueKt.createConstantValue(dataStream));
//        }
        return new CangJieAnnotationEntryStubImpl((StubElement<?>) parentStub, text, hasValueArguments );

    }
}

File: src/main/kotlin/com/huawei/cangjie/psi/stubs/elements/CjAnnotationEntryElementType.kt
package com.huawei.cangjie.psi.stubs.elements

import com.huawei.cangjie.psi.CjAnnotationEntry
import com.huawei.cangjie.psi.stubs.CangJieAnnotationEntryStub
import com.huawei.cangjie.psi.stubs.impl.CangJieAnnotationEntryStubImpl
import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import com.intellij.util.io.StringRef
import org.jetbrains.annotations.NonNls
import java.io.IOException

class CjAnnotationEntryElementType(debugName: @NonNls String) :
    CjStubElementType<CangJieAnnotationEntryStub , CjAnnotationEntry >(
        debugName,
        CjAnnotationEntry::class.java,
        CangJieAnnotationEntryStub::class.java
    ) {
    override fun createStub(
        psi: CjAnnotationEntry,
        parentStub: StubElement<out PsiElement?>
    ): CangJieAnnotationEntryStub {
        val shortName = psi.shortName
        val resultName = shortName?.asString()
        val valueArgumentList = psi.valueArgumentList
        val hasValueArguments = valueArgumentList != null && !valueArgumentList.arguments.isEmpty()
        return CangJieAnnotationEntryStubImpl(parentStub, StringRef.fromString(resultName), hasValueArguments)
    }

    @Throws(IOException::class)
    override fun serialize(stub: CangJieAnnotationEntryStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.getShortName())
        dataStream.writeBoolean(stub.hasValueArguments())
        if (stub is CangJieAnnotationEntryStubImpl) {
//            Map<Name, ConstantValue<?>> arguments = ((CangJieAnnotationEntryStubImpl) stub).getValueArguments();
//            dataStream.writeInt(arguments != null ? arguments.size() : 0);
//            if (arguments != null) {
//                for (Map.Entry<Name, ConstantValue<?>> valueEntry : arguments.entrySet()) {
//                    dataStream.writeName(valueEntry.getKey().asString());
//                    ConstantValue<?> value = valueEntry.getValue();
//                    CangJieConstantValueKt.serialize(value, dataStream);
//                }
//            }
        }
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>): CangJieAnnotationEntryStub {
        val text = dataStream.readName()
        val hasValueArguments = dataStream.readBoolean()
        //        int valueArgCount = dataStream.readInt();
//        Map<Name, ConstantValue<?>> args = new LinkedHashMap<>();
//        for (int i = 0; i < valueArgCount; i++) {
//            args.put(Name.identifier(Objects.requireNonNull(dataStream.readNameString())),
//                    CangJieConstantValueKt.createConstantValue(dataStream));
//        }
        return CangJieAnnotationEntryStubImpl(parentStub, text, hasValueArguments)
    }
}
